hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -1358,16 +1365,37 @@ private void verifyCapacityAndClasses(long capacitySize, String ioclass, String
   }
 
   private void parsePB(BucketCacheProtos.BucketCacheEntry proto) throws IOException {
+    backingMap = BucketProtoUtils.fromPB(proto.getDeserializersMap(), proto.getBackingMap(),
+      this::createRecycler);
+    prefetchCompleted.clear();
+    prefetchCompleted.putAll(proto.getPrefetchedFilesMap());
     if (proto.hasChecksum()) {
-      ((PersistentIOEngine) ioEngine).verifyFileIntegrity(proto.getChecksum().toByteArray(),
-        algorithm);
+      try {
+        ((PersistentIOEngine) ioEngine).verifyFileIntegrity(proto.getChecksum().toByteArray(),
+          algorithm);
+      } catch (IOException e) {
+        LOG.warn("Checksum for cache file failed. "
+          + "We need to validate each cache key in the backing map. This may take some time...");
+        long startTime = EnvironmentEdgeManager.currentTime();
+        int totalKeysOriginally = backingMap.size();
+        for (Map.Entry<BlockCacheKey, BucketEntry> keyEntry : backingMap.entrySet()) {
+          try {
+            ((FileIOEngine) ioEngine).checkCacheTime(keyEntry.getValue());
```

Comment:
This looks we are gonna check for each entry and if we have 30GB of backing Map for 2TB bucket cache. Then it could take a longer time, did we test it out with 2TB of bucket cache to determine how much time it will take to get the RS live?

If not performant, We should make this optional and verify the blocks when they are read or during the eviction scan.

Reply:
I have thought about it, as performance is not optimal (in my tests, 190GB cache with 2M blocks took around 105s to complete validation). The problem is that we have BucketAllocator, who needs to configure its indexes to buckets and offsets for the cache file according to the offset and length of each block in the backing map. If we don't do this validation, the backing map might be inconsistent, with more than one block for the same offset, then the bucket allocator initialisation will fail and we'll not able to access the blocks in the cache.

Reply:
On RS induced aborts, we sync the backing map when shutting down, so no need to validate the blocks when restarting and reloading the cache. I've done few tests where I've induced RS abort after HDFS/ZK errors during a write load that updates the cache, then on restart this validation wasn't needed (the backing map was consistent with the cache state). 

So I think the validation would happen only in cases where the RS crash abruptly, from a JVM fatal error or OS sigkill situation. For the record, I've tested crashing RS with a 1.6TB cache, the validation then took around 35 minutes to complete. This is not great, since the RS is useless during this time. One thought here is to allow bucket cache initialisation in the background, so that meanwhile reads would be going to the FS, but I would rather work on this in a separate jira/PR.

Reply:
@ankitsinghal , these add extra check to search for blocks in the bucket cache backing map first while prefetching, if we do find it there, we skip both the cache and/or file system reading altogether and increase the offset to check for the next block.